src/rgw/driver/d4n/rgw_sal_d4n.cc

```diff
@@ -1272,30 +1279,117 @@ int D4NFilterObject::D4NFilterDeleteOp::delete_obj(const DoutPrefixProvider* dpp
+    int ret = -1;
+    auto rgwObj = source->get_obj();
+    ret = source->get_obj_attrs(y, dpp, &rgwObj);
+    if (ret < 0) {
+      ldpp_dout(dpp, 0) << "D4NFilterObject::D4NFilterDeleteOp::" << __func__ << "(): failed to get obj attrs, ret=" << ret << dendl;
+      return next->delete_obj(dpp, y, flags);
+    }
+    
+    std::string version, policy_prefix;
+    if (source->have_instance()) {
+      version = source->get_instance();
+    } else {
+      version = source->get_object_version();
+    }
 
-  if ((ret = source->driver->get_cache_driver()->del(dpp, source->get_key().get_oid(), y)) < 0) 
-    ldpp_dout(dpp, 0) << "D4NFilterObject::" << __func__ << "(): CacheDriver del() method failed, ret=" << ret << dendl;
+    auto blockDir = source->driver->get_block_dir(); 
+    rgw::d4n::CacheBlock block;
+    block.cacheObj.objName = source->get_name();
+    block.cacheObj.bucketName = source->get_bucket()->get_name();
+    block.blockID = 0;
+    block.size = 0;
+
+    policy_prefix = head_oid_in_cache;
+    if (head_oid_in_cache.find("D_") != std::string::npos) {
+      policy_prefix = head_oid_in_cache.substr(2, head_oid_in_cache.size() - 2);  
+      obj.dirty = true;
+    }    
+
+    if ((ret = blockDir->del(dpp, &block, y)) == 0 || ret == -ENOENT) {
+      if ((ret = source->driver->get_cache_driver()->delete_data(dpp, head_oid_in_cache, y)) == 0) { // Sam: do we want del or delete_data here? 
+	if (!(ret = source->driver->get_policy_driver()->get_cache_policy()->erase(dpp, policy_prefix, y))) {
+	  ldpp_dout(dpp, 0) << "Failed to delete head policy entry for: " << obj.objName << ", ret=" << ret << dendl;
+	  return ret;
+        }
+      } else {
+	ldpp_dout(dpp, 0) << "Failed to delete head object for: " << obj.objName << ", ret=" << ret << dendl;
+	return ret;
+      }
+    } else {
+      ldpp_dout(dpp, 0) << "Failed to delete head object in block directory for: " << obj.objName << ", ret=" << ret << dendl;
+      return ret;
+    }
 
-  return next->delete_obj(dpp, y, flags);
-}
+    if ((ret = source->driver->get_obj_dir()->del(dpp, &obj, y)) < 0) {
+      ldpp_dout(dpp, 0) << "Failed to delete object directory entry for: " << obj.objName << ", ret=" << ret << dendl;
+      return ret;
+    }
+    
+    block.cacheObj.objName = source->get_key().get_oid();
+    off_t lst = source->get_obj_size();
+    off_t fst = 0;
+
+    do {
+      std::string prefix = source->get_bucket()->get_name() + "_" + version + "_" + obj.objName;
+      if (fst >= lst) {
+	break;
+      }
 
+      off_t cur_size = std::min<off_t>(fst + dpp->get_cct()->_conf->rgw_max_chunk_size, lst);
+      off_t cur_len = cur_size - fst;
+      block.blockID = static_cast<uint64_t>(fst);
+      block.size = static_cast<uint64_t>(cur_len);
+
+      if ((ret = blockDir->get(dpp, &block, y)) < 0) {
+	ldpp_dout(dpp, 0) << "Failed to retrieve directory entry for: " << obj.objName << " blockid: " << fst << " block size: " << cur_len << ", ret=" << ret << dendl;
+	return ret;
```

Comment:
did we not discuss that deletion of other blocks will continue in case there is an error for a previous block

Reply:
i remember now that we decided to return any error to the end user